### ISSUE TITLE
Method name shadows outer class used in the method signature

### ISSUE THREAD

**Issue by Andrej730** [astral-sh/ty#3581]
### Summary

Hi! Migrating from `pyright` to `ty`, stumbling upon function name shadowing the type available in the outer scope, see below - it results in an error and function returns `Unknown` instead of `declaration`.

```python
class declaration:
    ...

class A:
    # Function `declaration` is not valid in a return type annotation (invalid-type-form) [Ln 5, Col 30]
    def declaration(self) -> declaration: ...


# `a` is `Unknown`
a = A().declaration()
```

[Playground](https://play.ty.dev/95a1ea43-07bf-4bf3-a64b-bda66b8801b4)
Code sample in [pyright playground](https://pyright-play.net/?code=MYGwhgzhAEAmCmowCcwBcCWB7AdgLgChpjoA6cggpKaAQUJOgGJoAxAVx2E12gAMESVDxx9oGGDixpoANzAgMscTmhhoyeGnbJVaAJ4AHeGpxS06bKoAUGHPMWwAtAeNOAZlmQBbAJTQAbQAZVQBWABpoAGEsEGgAZgAGAF0iEgR3OERwYStrCHgQd38nAD4soUtcPDIKShY%2BMDEJfgBVHABrKQB3UQJ1AF46a19SQRyqnBGgA)

### Version

0.0.40

**Comment by tekumara** [comment]
Seeing this too, when a method has the same name as an imported type and `from __future__ import  annotations` is enabled.                                         

eg:

```python
from __future__ import annotations

from typing import Protocol
from datetime import date

class DateLike(Protocol):
    def date(self) -> date: ...
```

ty check:

```
   error[invalid-type-form]: Function `date` is not valid in a return type annotation                                                                
    --> repro.py:8:23                                                                                                                                
     |                                                                                                                                               
   8 |     def date(self) -> date: ...                                                                                                               
     |                       ^^^^                                                                                                                    
     |                                                                                                                                               
                                                                                                                                                     
   Found 1 diagnostic 
```

Doesn't happen in mypy or pyright.

**Comment by AlexWaygood** [comment]
Thanks for the report. This is #1747
